text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1159 @@
+  sections and has no effect on any existing schema functionality may be done
+  via incrementing the PATCH number only. This approach is only valid if the new
+  setting in the file is completely and safely ignorable by all existing
+  processing logic.
+
+  For example adding a completely new section that describes the full state of
+  the schema has no effect on existing consumers which only care about "changes"
+  section (unless we explicitly define the semantics of the new section such
+  that it _needs_ to be taken into account when processing schema changes). So,
+  adding such a new section can be done using a PATCH number increase.
+
+- MINOR number SHOULD be increased if a new setting is added to the file format
+  in a backward compatible manner. "Backward compatible" in this context means
+  that consumers that are aware of the new MINOR number can consume the file of
+  a particular MINOR version number or of any MINOR version number lower than
+  that, provided that MAJOR version numbers are the same. Typically, this means
+  that the added setting in file format is optional and the default value of the
+  setting matches the behavior of the previous file format version.
+
+  Note: there is no "forward compatibility" based on MINOR version number.
+  Consumers which support reading up to a particular MINOR version number SHOULD
+  NOT attempt to consume files with higher MINOR version numbers.
+
+- MAJOR number SHOULD be increased if the file format is changed in an
+  incompatible way. For example adding a new transformation type in the
+  "changes" section is an incompatible change because it cannot be ignored by
+  existing schema conversion logic, so such a change will require a new MAJOR
+  number.
+
+Correspondingly:
+
+- Consumers of the schema file SHOULD NOT attempt to interpret the schema file
+  if the MAJOR version number is different (higher or lower) than what the
+  consumer supports.
+
+- Consumers of the schema file SHOULD NOT attempt to interpret the schema file
+  if the MINOR version number is higher than what the consumer supports.
+
+- Consumers MAY ignore the PATCH number.
+
+To illustrate this with some examples:
+
+<table>
+  <tr>
+   <td><strong>File Format Version</strong>
+   </td>
+   <td><strong>Consumer's Expected Version</strong>
+   </td>
+   <td><strong>Consumer Can Read?</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>1.0.0
+   </td>
+   <td>1.0.0
+   </td>
+   <td>yes
+   </td>
+  </tr>
+  <tr>
+   <td>1.0.x
+   </td>
+   <td>1.0.y
+   </td>
+   <td>yes, for any x and y.
+   </td>
+  </tr>
+  <tr>
+   <td>1.a.x
+   </td>
+   <td>1.b.x
+   </td>
+   <td>if a&lt;b then yes, otherwise no.
+   </td>
+  </tr>
+  <tr>
+   <td>2.0.0
+   </td>
+   <td>1.x.y
+   </td>
+   <td>no
+   </td>
+  </tr>
+</table>
+
+## OTLP Changes
+
+To allow carrying the Schema URL in emitted telemetry it is necessary to add a
+schema_url field to OTLP messages.
+
+We add schema_url fields to the following messages:
```

Comment:
This is an interesting pattern, where we place a new optional field in one of two places, three times over.  Since it's likely to set a precedent, I have a couple of thought experiments: 

1. Instead of adding a new `schema_url` to the _ResourceSignal_ messages here, why not use a resource named `schema.url`? (JSON Schema uses `$schema`, I like that too). This would support schema URLS with no new fields.

2. Instead of adding a new `schema_url` to every of the _InstrumentationLibraryXXX_ messages, could we instead add another place for per-instrumentation-library resources to be expressed?  e.g., `opentelemetry.proto.resource.v1.Resource library_resource = 3;`. Would that be more flexible? I think it might be.

The reason for both of these questions is that I can imagine other kinds of property that are somehow special and if we add a new field for each one, we're in trouble. I'm thinking about information related to the metric collection interval and/or other kinds of processing that may have taken place as telemetry data is collected, information we may want to record either at the global level or at the per-library level.

In transportation systems, the term "Waybill" is used to document what is in transit in a multi-party exchange of goods. We have a similar situation here, where you want to express "what's inside" with a schema, and I strongly suspect there are other kinds of "what's inside" information ahead.

Reply:
I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.

Reply:
@jmacd 

>Instead of adding a new schema_url to the ResourceSignal messages here, why not use a resource named schema.url? (JSON Schema uses $schema, I like that too). This would support schema URLS with no new fields.

Do you mean a resource attribute named "schema.url"? That would work. The only thing I am not quite sure about is whether it is fair to say that "the telemetry schema is an attribute of a Resource that emits the telemetry".

>Instead of adding a new schema_url to every of the InstrumentationLibraryXXX messages, could we instead add another place for per-instrumentation-library resources to be expressed? e.g., opentelemetry.proto.resource.v1.Resource library_resource = 3;. Would that be more flexible? I think it might be.

Alternatively, we could allow `InstrumentationLibrary` to have arbitrary attributes and record "schema.url" attribute there. I am not quite sure recording a Resource inside an InstrumentationLibraryXXX message reflects our understanding of what a Resource is.

I don't have a strong opinion, open for suggestions.

Reply:
> I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.

@austinlparker size increase likely is negligible since the schema_url is recorded once per batch of telemetry emitted by instrumentation library.

Reply:
> > I'd also offer a non-specific concern over increasing the size of outgoing messages from OpenTelemetry, especially from things like the browser runtime. To that point, I'd support the idea of making the schema a named resource vs. adding fields.
> 
> @austinlparker size increase likely is negligible since the schema_url is recorded once per batch of telemetry emitted by instrumentation library.

Ah, ok, it read like you could apply multiple `schema.url` attributes. Looking at OTLP, it seems like adding this field to `InstrumentationLibrary` would make the most sense.

Reply:
>Maybe we just handle them as attributes rather than as resources and build semantic conventions up for them.

Yes, this is a possible option.

Reply:
I have put some more thought into this. While I believe it is a possible option I think it a tiny bit less desirable than using a distinct protocol field in OTLP. If the schema url is recorded simply as an attribute we need to come up with a name for the that attribute, let's assume for example it is `schema.url`. But what if we want to change the name of this attribute? We can't, this will be the one attribute which is not described and is not handled by the schema. This becomes an exception. I do not quite like having exceptions like this.
I think having a distinct new field in OTLP that is part of the proto is warranted in this case since it is a meta information that describes the attributes, it is not the same kind of information as all other attributes of a resource are.
So, for this reason my slight preference is to keep my original proposal.
To be clear: this does not prevent us from introducing attributes for instrumentation libraries. But that's a separate topic and can be done in a separate OTEP unrelated to this one.

Reply:
I agree.

Reply:
I concur with your logic (the "meta information that describes attributes" part). Perhaps you could add a note or errata that indicates why this information is special enough to warrant a new distinct field, in an attempt to keep this decision from establishing precedent? I think I'm more concerned about the potential for this decision to create a ripple effect, where we're introducing many versions of OTLP and causing frission for implementors and end-users alike.

Reply:
OK, Let me see if I can add something to the document.

To be clear: introducing `schema_url` as a new field in OTLP is a fully backwards compatible change. The default value for `schema_url` will be the empty string (as it normally works for new fields in Protobufs) which works great for our purposes. Empty `schema_url` indicates that the schema is unspecified (as you would expect from any pre-existing software that doesn't know about the concept of schema).
This is one of the fortunate cases when OTLP sender implementations and OTLP receiver implementations don't need to coordinate anything and can start supporting `schema_url` at their leisure. :-)